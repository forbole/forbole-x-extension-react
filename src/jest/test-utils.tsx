import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from '@mui/material';
import lightTheme from '../config/theme/lightTheme';
import '../assets/locales/i18n';

const AllTheProviders: FC = ({ children }) => {
  return <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
