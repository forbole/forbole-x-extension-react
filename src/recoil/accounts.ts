import { useCallback } from 'react'
import { atom, selectorFamily, selector, noWait, useSetRecoilState } from 'recoil'
import { fetchAccountBalance, fetchProfile } from '../fetches/accounts'
import chains from '../misc/chains'
import { passwordState } from './general'
import { encryptAndSaveToChromeStorage, removeStorage } from './utils/chromeStorageEncryption'

export const accountsState = atom<Account[]>({
  key: 'accounts',
  default: [],
  effects: [
    ({ onSet, getPromise }) => {
      onSet(async (newAccounts) => {
        const password = await getPromise(passwordState)
        if (password && newAccounts.length) {
          await encryptAndSaveToChromeStorage('accounts', newAccounts, password)
        } else {
          await removeStorage('accounts')
        }
      })
    },
  ],
})

export const accountState = selectorFamily<
  Account | undefined,
  { walletId: string; address: string }
>({
  key: 'account',
  get:
    ({ walletId, address }) =>
    ({ get }) =>
      get(accountsState).find((a) => a.walletId === walletId && a.address === address),
})

export const accountDetailState = selectorFamily<
  AccountDetail,
  { walletId: string; address: string }
>({
  key: 'accountDetail',
  get:
    ({ walletId, address }) =>
    async ({ get }) => {
      const account = get(accountState({ walletId, address }))
      const { balances, prices, delegations, unbonding, redelegations } = await fetchAccountBalance(
        account.chain,
        account.address
      )
      return { ...account, balances, prices, delegations, unbonding, redelegations }
    },
})

export const profileDetailState = selectorFamily<Profile, { walletId: string; address: string }>({
  key: 'profile',
  get:
    ({ walletId, address }) =>
    async ({ get }) => {
      const account = get(accountState({ walletId, address }))
      const response = await fetchProfile(account.chain, account.address)
      if (response.error) {
        throw response.error
      }
      return response
    },
})

export const walletAccountsState = selectorFamily<AccountDetail[], string>({
  key: 'walletAccounts',
  get:
    (walletId) =>
    ({ get }) => {
      const accounts = get(accountsState).filter((a) => a.walletId === walletId)
      const loadable = accounts.map((a) =>
        get(noWait(accountDetailState({ walletId: a.walletId, address: a.address })))
      )
      return accounts.map((a, i) =>
        loadable[i].state === 'hasValue'
          ? loadable[i].contents
          : {
              ...a,
              balances: { available: [], delegated: [], rewards: [], unbonding: [], total: [] },
              prices: [],
            }
      )
    },
})

export const useCreateAccounts = () => {
  const setAccounts = useSetRecoilState(accountsState)

  const createAccounts = useCallback(
    (params: { walletId: string; address: string; chain: string; hdPath: HdPath }[]) => {
      const createdAt = Date.now()
      setAccounts((accounts) => [
        ...accounts,
        ...params.map((account) => ({
          walletId: account.walletId,
          address: account.address,
          chain: account.chain,
          hdPath: {
            account: 0,
            change: 0,
            index: 0,
          },
          name: chains[account.chain].symbol,
          fav: false,
          createdAt,
        })),
      ])
    },
    [setAccounts]
  )

  return createAccounts
}

export const useChangeAccountName = () => {
  const setAccounts = useSetRecoilState(accountsState)

  const changeAccountName = useCallback(
    (params: { walletId: string; address: string; name: string }) => {
      setAccounts((accounts) =>
        accounts.map((a) =>
          a.walletId === params.walletId && a.address === params.address
            ? { ...a, name: params.name }
            : a
        )
      )
    },
    [setAccounts]
  )

  return changeAccountName
}

export const useDeleteAccount = () => {
  const setAccounts = useSetRecoilState(accountsState)

  const deleteAccount = useCallback(
    (params: { walletId: string; address: string }) => {
      setAccounts((accounts) =>
        accounts.filter((a) => !(a.walletId === params.walletId && a.address === params.address))
      )
    },
    [setAccounts]
  )

  return deleteAccount
}
