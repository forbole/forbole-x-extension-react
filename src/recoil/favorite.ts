import { atom, useSetRecoilState } from 'recoil';
import { getStorage, setStorage } from './utils/chromeStorageEncryption';
import { UpdatedAddress } from '../components/AddressBook/EditAddressDialog';

export type FavAddress = {
  address: string;
  crypto: string;
  moniker: string;
  note?: string;
  img?: string;
};

export const favAddressesState = atom<FavAddress[]>({
  key: 'favAddresses',
  default: (async () => {
    const favAddresses = await getStorage('favAddresses');
    return favAddresses ? JSON.parse(favAddresses) : [];
  })(),
  effects: [
    ({ onSet }) => {
      onSet((addresses) => {
        const favAddresses = JSON.stringify(addresses);
        setStorage({ favAddresses });
      });
    },
  ],
});

export const useFavAddressesMutations = () => {
  const setFavAddresses = useSetRecoilState(favAddressesState);

  const addFavAddresses = (newAddress: FavAddress) => {
    setFavAddresses((vs) => {
      if (vs.find((e) => e.address === newAddress.address)) {
        return vs;
      }
      return [newAddress, ...vs];
    });
  };

  const updateFavAddresses = (updatedAddress: UpdatedAddress) => {
    setFavAddresses((vs) =>
      vs.map((add) => {
        if (add.address === updatedAddress.address) {
          return { ...updatedAddress, address: updatedAddress.newAddress, newAddress: null };
        }
        return add;
      })
    );
  };

  const deleteFavAddresses = (targetAddress: string) => {
    setFavAddresses((vs) =>
      vs.filter((add) => {
        return add.address !== targetAddress;
      })
    );
  };

  return { addFavAddresses, updateFavAddresses, deleteFavAddresses };
};
