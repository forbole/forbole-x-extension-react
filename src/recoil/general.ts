import { atom, useSetRecoilState } from 'recoil'
import { useCallback } from 'react'
import nightwind from 'nightwind/helper'
import { decryptChromeStorage, getStorage } from './utils/chromeStorageEncryption'
import { walletsState } from './wallets'
import { accountsState } from './accounts'

export const themeState = atom<string>({
  key: 'theme',
  default: (async () => {
    const theme = await getStorage('nightwind-mode')
    return theme ?? 'light'
  })(),
})

export const passwordState = atom<string>({
  key: 'password',
  default: '',
})

export const isFirstTimeUserState = atom<boolean>({
  key: 'firsTimeUser',
  default: (async () => {
    const walletString = await getStorage('wallets')
    return !walletString
  })(),
})

export const useSetTheme = () => {
  const setThemeState = useSetRecoilState(themeState)

  const setTheme = useCallback(async () => {
    nightwind.toggle()
    const theme = await getStorage('nightwind-mode')
    setThemeState(theme)
  }, [setThemeState])

  return setTheme
}

export const useCreatePassword = () => {
  const setPassword = useSetRecoilState(passwordState)

  const createPassword = useCallback(
    (pw: string) => {
      setPassword(pw)
    },
    [setPassword]
  )

  return createPassword
}

export const useUpdatePassword = () => {
  const setPassword = useSetRecoilState(passwordState)
  const setWallets = useSetRecoilState(walletsState)
  const setAccounts = useSetRecoilState(accountsState)

  const updatePassword = useCallback(
    async (oldPw: string, newPw: string) => {
      const decryptedWallets = await decryptChromeStorage<Wallet[]>('wallets', oldPw)
      const decryptedAccounts = await decryptChromeStorage<Account[]>('accounts', oldPw)

      setPassword(newPw)
      setWallets(decryptedWallets)
      setAccounts(decryptedAccounts)
    },
    [setPassword, setWallets, setAccounts]
  )

  return updatePassword
}

export const useUnlockWallets = () => {
  const setPassword = useSetRecoilState(passwordState)
  const setWallets = useSetRecoilState(walletsState)
  const setAccounts = useSetRecoilState(accountsState)

  const unlockWallets = useCallback(
    async (pw: string) => {
      console.log(pw)
      const decryptedWallets = await decryptChromeStorage<Wallet[]>('wallets', pw)
      const decryptedAccounts = await decryptChromeStorage<Account[]>('accounts', pw)
      setPassword(pw)
      setWallets(decryptedWallets)
      setAccounts(decryptedAccounts)
    },
    [setWallets, setAccounts, setPassword]
  )

  return unlockWallets
}
