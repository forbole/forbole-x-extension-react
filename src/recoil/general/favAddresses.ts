import { atom, useRecoilState } from "recoil";

type FavAddress = {
  address: string;
  crypto: string;
  moniker: string;
  note?: string;
  img?: string;
};

export const favAddressesState = atom({
  key: "favAddresses",
  default: [] as FavAddress[],
});

export const useFavAddresses = () => {
  const [favAddresses, setFavAddresses] = useRecoilState(favAddressesState);

  const update = (addresses: FavAddress[]) => {
    setFavAddresses(addresses);
  };

  const addFavAddress = (address: FavAddress) => {
    const exist = favAddresses.some((v) => {
      v.address === address.address;
    });
    if (!exist) {
      const newValidatorsList = [...favAddresses, address];
      update(newValidatorsList);
    }
  };

  return { favAddresses, addFavAddress };
};
