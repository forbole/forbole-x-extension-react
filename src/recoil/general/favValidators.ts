import { useCallback } from "react";
import { atom, useRecoilState } from "recoil";

type Validators = string[];

export const favValidatorsState = atom({
  key: "favValidators",
  default: [] as Validators,
});

export const useFavValidators = () => {
  const [favValidators, setFavValidators] = useRecoilState(favValidatorsState);

  const update = (validators: Validators) => {
    setFavValidators(validators);
  };

  const addValidator = (validator: string) => {
    const exist = favValidators.some((v) => {
      v === validator;
    });
    if (!exist) {
      const newValidatorsList = [...favValidators, validator];
      update(newValidatorsList);
    }
  };

  return { favValidators, addValidator };
};
