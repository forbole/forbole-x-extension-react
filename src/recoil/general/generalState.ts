import { useCallback } from "react";
import { atom, useRecoilState } from "recoil";
import currencies from "../../misc/currencies";

export const alwaysRequirePasswordState = atom({
  key: "alwaysRequirePassword",
  default: false as boolean,
});

export const currencyState = atom({
  key: "currency",
  default: "USD" as typeof currencies[number],
});

export const useAlwaysRequirePassword = () => {
  const [alwaysRequirePassword, setAlwaysRequirePassword] = useRecoilState(
    alwaysRequirePasswordState
  );

  const set = useCallback(() => {
    setAlwaysRequirePassword(!alwaysRequirePassword);
  }, [alwaysRequirePassword, setAlwaysRequirePassword]);

  return [alwaysRequirePassword, set] as const;
};
