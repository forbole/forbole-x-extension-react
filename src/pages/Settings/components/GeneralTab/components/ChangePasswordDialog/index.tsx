import React from 'react';
import { Box, DialogContent, DialogActions, Typography, Button } from '@mui/material';
import { useTranslation } from 'react-i18next';
import Layout from 'components/Layout/layout';
import TxSuccess from 'components/svg/TxSuccess';
import IconCross from 'components/svg/IconCross';
import { useUnlockWallets, useUpdatePassword } from '../../../../../../recoil/general';
import PasswordInput from '../../../../../../components/inputs/PasswordInput';

type Props = {
  /**
   * What to do when the dialog is closed
   */
  onClose?: () => void;
};
// false positive
// eslint-disable-next-line no-shadow
enum DialogStages {
  ENTER_CURRENT_PW = 0,

  ENTER_NEW_PW = 1,

  SUCCESS = 2,
}

const ChangePasswordDialog = ({ onClose }: Props) => {
  const { t } = useTranslation('settings');

  const [stage, setStage] = React.useState<DialogStages>(DialogStages.ENTER_CURRENT_PW);

  const [oldPassword, setOldPassword] = React.useState('');

  const [newPassword, setNewPassword] = React.useState('');

  const [error, setError] = React.useState('');

  const unlockWallet = useUnlockWallets();

  const updatePassword = useUpdatePassword();

  React.useEffect(() => {
    // reset error on stage change
    setError('');
  }, [stage]);

  const resetDialog = React.useCallback(() => {
    setStage(DialogStages.ENTER_CURRENT_PW);
    setError('');
    setOldPassword('');
    setNewPassword('');
    onClose();
  }, []);

  const onSubmit = React.useCallback(async () => {
    if (stage === DialogStages.ENTER_CURRENT_PW) {
      try {
        await unlockWallet(oldPassword);
        setStage(DialogStages.ENTER_NEW_PW);
      } catch (err) {
        setError(t('general.changePwDialog.errorIncorrect'));
      }
    } else if (stage === DialogStages.ENTER_NEW_PW) {
      if (newPassword.length < 6) {
        setError(t('general.changePwDialog.error6Char'));
      } else {
        try {
          await updatePassword(oldPassword, newPassword);
          setStage(DialogStages.SUCCESS);
        } catch (err) {
          console.log(err);
        }
      }
    }
  }, [stage, oldPassword, newPassword]);

  if (stage === DialogStages.SUCCESS)
    return (
      <Layout
        hideLeftElement
        rightElement={
          <Button
            disableElevation
            disableRipple
            onClick={resetDialog}
            sx={{
              backgroundColor: 'transparent',
              padding: 0,
              margin: 0,
              minWidth: 0,
              borderWidth: 0,
              '&:hover': {
                background: 'none',
              },
            }}
          >
            <IconCross />
          </Button>
        }
      >
        <Box
          sx={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
          }}
        >
          <TxSuccess />
          <Typography variant="h5" textAlign="center" marginBottom={4}>
            {t('general.changePwDialog.stage3.title')}
          </Typography>

          <Typography variant="subtitle2">
            {t('general.changePwDialog.stage3.description')}
          </Typography>
        </Box>
      </Layout>
    );

  return (
    <Layout title="" backCallback={resetDialog}>
      <form
        noValidate
        onSubmit={(e) => {
          e.preventDefault();
          onSubmit();
        }}
      >
        <Typography variant="h3" textAlign="center" marginBottom={4}>
          {t('general.changePwDialog.stage1.title')}
        </Typography>
        <DialogContent>
          <Box mb={18}>
            <Typography gutterBottom>
              {t(
                stage === DialogStages.ENTER_CURRENT_PW
                  ? 'general.changePwDialog.stage1.label'
                  : 'general.changePwDialog.stage2.label'
              )}
            </Typography>
            {stage === DialogStages.ENTER_CURRENT_PW ? (
              <PasswordInput
                value={oldPassword}
                onChange={(e) => setOldPassword(e.target.value)}
                placeholder={t('general.changePwDialog.stage1.placeholder')}
                error={!!error}
                helperText={error}
              />
            ) : (
              <PasswordInput
                value={newPassword}
                onChange={(e) => setNewPassword(e.target.value)}
                placeholder={t('general.changePwDialog.stage2.placeholder')}
                error={!!error}
                helperText={error}
                withSecurityLevel
              />
            )}
          </Box>
        </DialogContent>
        <DialogActions>
          <Button
            sx={(theme) => ({ margin: theme.spacing(4, 2), color: 'white' })}
            fullWidth
            variant="contained"
            color="primary"
            type="submit"
          >
            {t(stage === DialogStages.ENTER_NEW_PW ? 'save' : 'next')}
          </Button>
        </DialogActions>
      </form>
    </Layout>
  );
};

export default ChangePasswordDialog;
