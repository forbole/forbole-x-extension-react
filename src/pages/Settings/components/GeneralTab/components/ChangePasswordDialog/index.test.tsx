import React from 'react'
import i18next from 'i18next'
import { fireEvent, waitFor } from '@testing-library/dom'
import ChangePasswordDialog from './index'
import { render } from '../../../../../../jest/test-utils'
import { useUnlockWallets, useUpdatePassword } from '../../../../../../recoil/general'

jest.mock('../../../../../../recoil/general', () => ({
  useUnlockWallets: jest.fn(),
  useUpdatePassword: jest.fn(),
}))
// jest.mock('../../../../../../recoil/general/useUpdatePassword', () => true)

describe('component: ChangePasswordDialog', () => {
  it('renders', () => {
    const t = render(<ChangePasswordDialog isOpen />)

    expect(t).toMatchSnapshot()
  })

  it('goes from stage 1 -> 2', async () => {
    ;(useUnlockWallets as jest.Mock).mockReturnValueOnce(() => async () => true)
    ;(useUpdatePassword as jest.Mock).mockReturnValueOnce(() => () => true)

    const { getByText, getByPlaceholderText } = render(<ChangePasswordDialog isOpen />)

    // expect stage 1 input label
    expect(getByText(i18next.t('settings:general.changePwDialog.stage1.label') as string))

    const stage1Input = getByPlaceholderText(
      i18next.t('settings:general.changePwDialog.stage1.placeholder') as string
    )

    fireEvent.change(stage1Input, 'abc123')
    fireEvent.click(getByText(i18next.t('settings:next') as string))

    await waitFor(() => {
      expect(getByText(i18next.t('settings:general.changePwDialog.stage2.label') as string))
    })
  })

  // not sure why this doesn't work
  // it('goes from stage 2 -> 3', async () => {
  //   ;(useUnlockWallets as jest.Mock).mockReturnValueOnce(() => async () => true)
  //   ;(useUpdatePassword as jest.Mock).mockReturnValueOnce(() => async () => true)
  //
  //   const { getByText, getByPlaceholderText } = render(<ChangePasswordDialog isOpen />)
  //
  //   const stage1Input = getByPlaceholderText(
  //     i18next.t('settings:general.changePwDialog.stage1.placeholder') as string
  //   )
  //
  //   fireEvent.change(stage1Input, 'abc123')
  //   fireEvent.click(getByText(i18next.t('settings:next') as string))
  //
  //   await waitFor(() => {
  //     expect(getByText(i18next.t('settings:general.changePwDialog.stage2.label') as string))
  //   })
  //
  //   const stage2Input = getByPlaceholderText(
  //     i18next.t('settings:general.changePwDialog.stage2.placeholder') as string
  //   )
  //
  //   fireEvent.change(stage2Input, 'abc1231')
  //   fireEvent.click(getByText(i18next.t('settings:save') as string))
  //
  //   await waitFor(() => {
  //     expect(getByText(i18next.t('settings:general.changePwDialog.stage3.description') as string))
  //   })
  // })
})
