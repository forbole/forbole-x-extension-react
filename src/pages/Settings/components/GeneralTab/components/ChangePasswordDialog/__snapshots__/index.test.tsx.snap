// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ChangePasswordDialog renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiModal-root MuiDialog-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby="mui-1"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-n2ob6u-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <form
            novalidate=""
          >
            <h2
              class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-1f9hik6-MuiTypography-root-MuiDialogTitle-root"
              id="mui-1"
            >
              Change Unlock Password
            </h2>
            <div
              class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
            >
              <div
                class="MuiBox-root css-xsyq4r"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-szuhpm-MuiTypography-root"
                >
                  Enter current unlock password
                </p>
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root outline-0 css-wb57ya-MuiFormControl-root-MuiTextField-root"
                >
                  <div
                    class="MuiFilledInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-1o5541w-MuiInputBase-root-MuiFilledInput-root"
                  >
                    <input
                      aria-invalid="false"
                      autocomplete="one-time-code"
                      class="MuiFilledInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-pecu04-MuiInputBase-input-MuiFilledInput-input"
                      id="mui-2"
                      placeholder="Password"
                      type="password"
                      value=""
                    />
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-jv61v3-MuiInputAdornment-root"
                    >
                      <button
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                        tabindex="0"
                        type="button"
                      >
                        <svg
                          fill="#8B8B8B"
                          height="24px"
                          width="24px"
                        >
                          icon_hide.svg
                        </svg>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  
                </div>
              </div>
            </div>
            <div
              class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-n0bxl-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="submit"
              >
                Next
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
